### PR TITLE
Override Response.setTimeout to make response timeout handler work

### PR DESCRIPTION
**Pull request by gyszalai**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/response.js [hunk 1]
```diff
@@ -46,7 +46,18 @@ function Response (req, onEnd, reject) {
 
 util.inherits(Response, http.ServerResponse)
 
+let timeoutHandle
+Response.prototype.setTimeout = function (msecs, callback) {
+  if (callback) {
+    timeoutHandle = setTimeout(callback, msecs)
```
Comment (mcollina) [inline lib/response.js]: We will need to emit the `'timeout'` event when the setTimeout triggers (independently of the presence of the callback)

See https://nodejs.org/api/net.html#net_socket_settimeout_timeout_callback for more details.

File: lib/response.js [hunk 2]
```diff
@@ -46,7 +46,18 @@ function Response (req, onEnd, reject) {
 
 util.inherits(Response, http.ServerResponse)
 
+let timeoutHandle
```
[the remark this answers is not in the document]
Comment (gyszalai) [reply lib/response.js]: @Eomm Id don't really get it. do you mean I should set it as a property and access it as `this.timeoutHandle`?

File: lib/response.js [hunk 3]
```diff
@@ -46,7 +46,18 @@ function Response (req, onEnd, reject) {
...
+Response.prototype.setTimeout = function (msecs, callback) {
+  if (callback) {
```
[the remark this answers is not in the document]
Comment (gyszalai) [reply lib/response.js]: ok, shall I throw an error if no callback is specified?

File: lib/response.js (hunk 1 above)
Comment (gyszalai) [reply lib/response.js]: @mcollina according to the docs, the timeout event should be emitted by the socket. how can I cet a reference to the underlying socket?

Comment (gyszalai) [reply lib/response.js]: is calling `this.socket.emit('timeout')` enough?

File: lib/response.js [hunk 4]
```diff
@@ -46,7 +46,18 @@ function Response (req, onEnd, reject) {
...
+    timeoutHandle = setTimeout(callback, msecs)
+  }
+  return this
+}
+
 Response.prototype.writeHead = function () {
+  if (timeoutHandle) {
+    clearTimeout(timeoutHandle)
```
[the remark this answers is not in the document]
Comment (gyszalai) [reply lib/response.js]: @Eomm ok, I'll move it there

File: lib/response.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gyszalai) [reply lib/response.js]: According to the http docs, the callback is optional: https://nodejs.org/api/http.html#http_response_settimeout_msecs_callback

File: lib/response.js [hunk 5]
```diff
@@ -46,6 +46,16 @@ function Response (req, onEnd, reject) {
 
 util.inherits(Response, http.ServerResponse)
 
+Response.prototype.setTimeout = function (msecs, callback) {
+  this.timeoutHandle = setTimeout(() => {
+    this.socket.emit('timeout')
```
Comment (mcollina) [inline lib/response.js]: It should be `this.emit('timeout')`.

File: lib/response.js [hunk 6]
```diff
@@ -46,6 +46,16 @@ function Response (req, onEnd, reject) {
...
+    if (callback) {
+      callback()
+    }
```
Comment (mcollina) [inline lib/response.js]: I would replace this with `this.on('timeout', callback)` outside of the setTimeout callback

Comment (gyszalai) [reply lib/response.js]: done

File: lib/response.js (hunk 5 above)
Comment (gyszalai) [reply lib/response.js]: done

**Comment by gyszalai** [comment]
Somebody please, merge this PR.

**Pull request merged**